### PR TITLE
Homework 3

### PR DESCRIPTION
**Pull request by citruscaviar**

### CODE REVIEW

File: src/main/java/com/onlinebookstore/store/controller/AuthController.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+@Tag(name = "Authentication management",description = "Manages authentication controller")
+@RestController
+@RequiredArgsConstructor
+@RequestMapping(value = "/auth")
+public class AuthController {
+    private final UserService userService;
+
+    @PostMapping("/registration")
+    @Operation()
```
Comment (liliia-ponomarenko) [inline src/main/java/com/onlinebookstore/store/controller/AuthController.java:23]: Let's add a description to the Operation annotation

File: src/main/java/com/onlinebookstore/store/controller/AuthController.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    public UserResponseDto register(
+            @RequestBody UserRegistrationRequestDto requestDto)
```
Comment (liliia-ponomarenko) [inline src/main/java/com/onlinebookstore/store/controller/AuthController.java:25]: Valid annotation missed

File: src/main/java/com/onlinebookstore/store/service/impl/UserServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+@Service
+@RequiredArgsConstructor
+public class UserServiceImpl implements UserService {
+    private final UserRepository userRepository;
+    private final UserMapper userMapper;
+
+    @Override
+    public UserResponseDto register(
+            UserRegistrationRequestDto
+            requestDto) throws RuntimeException {
```
Comment (liliia-ponomarenko) [inline src/main/java/com/onlinebookstore/store/service/impl/UserServiceImpl.java:22]: why do you need to throw RuntimeException here?

File: src/main/java/com/onlinebookstore/store/service/impl/UserServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        if (userRepository.findByEmail(requestDto.getEmail()).isPresent()) {
```
Comment (liliia-ponomarenko) [inline src/main/java/com/onlinebookstore/store/service/impl/UserServiceImpl.java:23]: better use `existByEmail`

File: src/main/java/com/onlinebookstore/store/service/impl/UserServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+            throw new RegistrationException("Unable to complete registration");
```
Comment (liliia-ponomarenko) [inline src/main/java/com/onlinebookstore/store/service/impl/UserServiceImpl.java:24]: Let's make the message more informative (with reason)

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution ;)

File: src/main/java/com/onlinebookstore/store/controller/AuthController.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+import lombok.RequiredArgsConstructor;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@Tag(name = "Authentication management",description = "Manages authentication controller")
+@RestController
+@RequiredArgsConstructor
+@RequestMapping(value = "/auth")
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/store/controller/AuthController.java:18]:
```suggestion
@RequestMapping("/auth")
```

File: src/main/java/com/onlinebookstore/store/exceptions/GlobalExceptionHandler.java [hunk 7]
```diff
@@ -23,6 +23,11 @@ public ResponseEntity<List<String>> handleValidationExceptions(
 
     @ExceptionHandler(EntityNotFoundException.class)
     public ResponseEntity<String> handleEntityNotFoundExceptions(EntityNotFoundException ex) {
-        return new ResponseEntity<>("EntityNotFoundException occured",HttpStatus.NOT_FOUND);
+        return new ResponseEntity<>("EntityNotFoundException has occurred",HttpStatus.NOT_FOUND);
+    }
+
+    @ExceptionHandler(RegistrationException.class)
+    public ResponseEntity<String> handleRegistrationExceptions(RegistrationException ex) {
+        return new ResponseEntity<>("RegistrationException has occurred",HttpStatus.BAD_REQUEST);
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/store/exceptions/GlobalExceptionHandler.java:31]:
```suggestion
        return new ResponseEntity<>("RegistrationException has occurred",HttpStatus.CONFLICT);
```

File: src/main/java/com/onlinebookstore/store/model/User.java [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+@Getter
+@Setter
+@SQLDelete(sql = "UPDATE users SET is_deleted = true WHERE id = ?")
+@SQLRestriction("is_deleted = false")
+@Table(name = "users")
+public class User {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Column(nullable = false)
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/store/model/User.java:24]:
```suggestion
    @Column(nullable = false, unique = true
```

File: src/main/java/com/onlinebookstore/store/service/UserService.java [hunk 9]
```diff
@@ -0,0 +1,12 @@
+package com.onlinebookstore.store.service;
+
+import com.onlinebookstore.store.dto.UserRegistrationRequestDto;
+import com.onlinebookstore.store.dto.UserResponseDto;
+import com.onlinebookstore.store.exceptions.RegistrationException;
+
+public interface UserService {
+
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/store/service/UserService.java:8]:
```suggestion
```

File: src/main/java/com/onlinebookstore/store/repository/UserRepository.java [hunk 10]
```diff
@@ -0,0 +1,9 @@
+package com.onlinebookstore.store.repository;
+
+import com.onlinebookstore.store.model.User;
+import java.util.Optional;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface UserRepository extends JpaRepository<User, Long> {
+    Optional<User> findByEmail(String email);
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/store/repository/UserRepository.java:8]:
```suggestion
    boolean existsByEmail(String email);
```

File: src/main/java/com/onlinebookstore/store/service/impl/UserServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+@Service
+@RequiredArgsConstructor
+public class UserServiceImpl implements UserService {
+    private final UserRepository userRepository;
+    private final UserMapper userMapper;
+
+    @Override
+    public UserResponseDto register(
+            UserRegistrationRequestDto
+            requestDto) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/store/service/impl/UserServiceImpl.java:22]:
```suggestion
    public UserResponseDto register(UserRegistrationRequestDto requestDto) {
```

File: src/main/java/com/onlinebookstore/store/service/impl/UserServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+        if (userRepository.findByEmail(requestDto.getEmail()).isPresent()) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/store/service/impl/UserServiceImpl.java:23]:
```suggestion
        if (userRepository.existsByEmail(request.getEmail())) {
```

File: src/main/java/com/onlinebookstore/store/service/impl/UserServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+            throw new RegistrationException("Unable to complete registration: email "
+                + requestDto.getEmail() + " is already in use.");
+        }
+
+        User user = userMapper.toModel(requestDto);
+
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/store/service/impl/UserServiceImpl.java:29]:
```suggestion
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

**Review by Elena-Bruyako — APPROVED** [review]

**Pull request merged**